allennlp/commands/predict.py

```diff
@@ -94,62 +102,95 @@ def _get_predictor(args: argparse.Namespace) -> Predictor:
 
     return Predictor.from_archive(archive, args.predictor)
 
-def _run(predictor: Predictor,
-         input_file: IO,
-         output_file: Optional[IO],
-         batch_size: int,
-         print_to_console: bool) -> None:
 
-    def _run_predictor(batch_data):
+class _Predict:
+
+    def __init__(self,
+                 predictor: Predictor,
+                 input_file: str,
+                 output_file: Optional[str],
+                 batch_size: int,
+                 print_to_console: bool,
+                 has_dataset_reader: bool) -> None:
+
+        self._predictor = predictor
+        self._input_file = input_file
+        if output_file is not None:
+            self._output_file = open(output_file, "w")
+        else:
+            self._output_file = None
+        self._batch_size = batch_size
+        self._print_to_console = print_to_console
+        if has_dataset_reader:
+            self._dataset_reader = predictor._dataset_reader # pylint: disable=protected-access
+        else:
+            self._dataset_reader = None
+
+    def _predict_json(self, batch_data: List[JsonDict]) -> Iterator[str]:
         if len(batch_data) == 1:
-            result = predictor.predict_json(batch_data[0])
-            # Batch results return a list of json objects, so in
-            # order to iterate over the result below we wrap this in a list.
-            results = [result]
+            result = self._predictor.predict_json(batch_data[0])
+            yield self._predictor.dump_line(result)
         else:
-            results = predictor.predict_batch_json(batch_data)
+            results = self._predictor.predict_batch_json(batch_data)
+        for output in results:
+            yield self._predictor.dump_line(output)
```

Comment:
this might be marginally cleaner as

```
if len(batch_data) == 1:
    results = [self._predictor.predict_json(batch_data[0])]
else:
    results = self._predictor.predict_batch_json(batch_data)
for output in results:
    ...
```

what do you think?